src/main/java/com/minecolonies/coremod/entity/ai/citizen/herders/AbstractEntityAIHerder.java

```diff
@@ -0,0 +1,420 @@
+          new AITarget(PREPARING, this::prepareForHerding),
+          new AITarget(HERDER_DECIDE, this::decideWhatToDo),
+          new AITarget(HERDER_BREED, this::breedAnimals),
+          new AITarget(HERDER_BUTCHER, this::butcherAnimals)
+        );
+        worker.setCanPickUpLoot(true);
+    }
+
+    @Override
+    protected int getActionsDoneUntilDumping()
+    {
+        return 10;
+    }
+
+    /**
+     * Decides what job the herder should switch to, breeding, butchering, or shearing (If Shepherd).
+     *
+     * @return the next AIState the shepherd should switch to, after executing this method.
+     */
+    public AIState decideWhatToDo()
+    {
+        setDelay(40);
+
+        final List<T> animals = new ArrayList<>(getAnimals());
+
+        if (animals.isEmpty())
+        {
+            setDelay(100);
+            return HERDER_DECIDE;
+        }
+
+        worker.setLatestStatus(new TextComponentTranslation("com.minecolonies.coremod.status.herder.deciding"));
+
+        final int numOfBreedableAnimals = animals.stream().filter(animal -> animal.getGrowingAge() == 0).toArray().length;
+
+        if (maxAnimals())
+        {
+            return HERDER_BUTCHER;
+        }
+        else if (numOfBreedableAnimals >= NUM_OF_ANIMALS_TO_BREED)
+        {
+            return HERDER_BREED;
+        }
+        return HERDER_DECIDE;
+    }
+
+    /**
+     * Redirects the herder to their building.
+     *
+     * @return the next state.
+     */
+    private AIState startWorkingAtOwnBuilding()
+    {
+        worker.setLatestStatus(new TextComponentTranslation("com.minecolonies.coremod.status.herder.goingToHut"));
+        if (walkToBuilding())
+        {
+            return getState();
+        }
+        return PREPARING;
+    }
+
+    /**
+     * Prepares the herder for herding
+     *
+     * @return the next AIState
+     */
+    private AIState prepareForHerding()
+    {
+        if (checkForToolOrWeapon(ToolType.AXE) && checkOrRequestItems(new ItemStack(Items.WHEAT, 2))
+              && (this instanceof EntityAIWorkShepherd && checkForToolOrWeapon(ToolType.SHEARS)))
+        {
+            return getState();
+        }
+        return HERDER_DECIDE;
+    }
+
+    /**
+     * Butcher some animals (Preferably Adults) that the herder looks after.
+     *
+     * @return The next AIState
+     */
+    private AIState butcherAnimals()
+    {
+        setDelay(20);
+
+        if (!maxAnimals())
+        {
+            return HERDER_DECIDE;
+        }
+
+        if (!equipTool(ToolType.AXE))
+        {
+            return PREPARING;
+        }
+
+        EntityAnimal animal = getAnimals().stream().filter(animalToButcher -> !animalToButcher.isChild()).findFirst().orElse(null);
+
+        if (animal == null)
+        {
+            animal = getAnimals().stream().findFirst().orElse(null);
+        }
+
+        butcherAnimal(animal);
+
+        if (!animal.isEntityAlive())
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') A "NullPointerException" could be thrown; "animal" is nullable here. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2259)